### PR TITLE
fix(sink): derive append only from optimizer when no format declared

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Add a branch for the case that the user does not declare sink type/sink format,
- allow force append only when there is no FORMAT clause

 which is especially useful for "CREATE SINK INTO TABLE"

```SQL
create table t(a int) append only;
create table tt(a int) append only;
create table ttt(a int);

dev=> 
explain create sink s into t as select a from tt;
                             QUERY PLAN                              
---------------------------------------------------------------------
 StreamProject { exprs: [tt.a, null:Serial] }
 └─StreamSink { type: AppendOnly, columns: [a, tt._row_id(hidden)] }
   └─StreamTableScan { table: tt, columns: [a, _row_id] }
(3 rows)

dev=> 
explain create sink s into t as select a from ttt with (force_append_only = true);
                                QUERY PLAN                                 
---------------------------------------------------------------------------
 StreamProject { exprs: [ttt.a, null:Serial] }
 └─StreamSink { type: ForceAppendOnly, columns: [a, ttt._row_id(hidden)] }
   └─StreamTableScan { table: ttt, columns: [a, _row_id] }
(3 rows)
```

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by xiangjinwu — APPROVED** [review]
Code itself lgtm as long as all the behavior changes are intentional. Please confirm before merge.

File: src/frontend/src/optimizer/plan_node/stream_sink.rs
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:297]: Thanks for your clarification. If we do not maintain the compatibility for those errored statements, the only change is this one.

> this is a backward incompatible change: NotDefined was considered false before this PR, even when frontend derives it is actually append only.

Comment (st1page) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:297]: Thanks for your clarification. If we do not maintain the compatibility for those errored statements, the only change is this one.

> this is a backward incompatible change: NotDefined was considered false before this PR, even when frontend derives it is actually append only.

Comment (xiangjinwu) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:297]: > If we do not maintain the compatibility for those errored statements

In my opinion, it is okay now to move from err to non-err in this PR. However, if we want to move from non-err to err again in the future, it would be a breaking change. This is what I mean by *making the backward compatibility promise harder to maintain*.

**Comment by st1page** [comment]
After offline, it will be implemented with a less invasive method

**Pull request closed without merging**
